lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.reef.driver.context.ActiveContext;
+import org.apache.reef.driver.context.ClosedContext;
+import org.apache.reef.driver.context.ContextBase;
+import org.apache.reef.tang.ClassHierarchy;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Factory for ActiveContextBridge instances.
+ * <p/>
+ * This class is thread safe because all its methods are synchronized. It should only be used within the bridge code.
+ */
+@DriverSide
+@ThreadSafe
+@Private
+public final class ActiveContextBridgeFactory {
+  private static final Logger LOG = Logger.getLogger(ActiveContextBridge.class.getName());
+
+  private final Map<String, ActiveContextBridge> activeContextBridgeCache = new HashMap<>();
+  private final ClassHierarchy clrClassHierarchy = Utilities.loadClassHierarchy(NativeInterop.CLASS_HIERARCHY_FILENAME);
+  private final AvroConfigurationSerializer configurationSerializer;
+
+  /**
+   * This is always instantiated via Tang.
+   *
+   * @param configurationSerializer passed to the ActiveContextBridge instances for configuration serialization.
+   */
+  @Inject
+  private ActiveContextBridgeFactory(final AvroConfigurationSerializer configurationSerializer) {
+    this.configurationSerializer = configurationSerializer;
+  }
+
+  /**
+   * Instantiates a new ActiveContextBridge or returns the one created earlier.
+   *
+   * @param context the context for which to return an ActiveContextBridge instance.
+   * @return a new ActiveContextBridge or returns the one created earlier.
+   */
+  public synchronized ActiveContextBridge getActiveContextBridge(final ActiveContext context) {
+    final String contextId = getUniqueContextId(context);
+
+    ActiveContextBridge result = activeContextBridgeCache.get(contextId);
+    if (null == result) {
+      result = new ActiveContextBridge(context, this.clrClassHierarchy, this.configurationSerializer);
+      activeContextBridgeCache.put(contextId, result);
+    }
+    return result;
+  }
+
+  /**
+   * Removes the ActiveContextBridge associated with the given Context from the cache.
+   *
+   * @param context the context whose ActiveContextBridge instance shall be removed from the cache.
+   */
+  public synchronized void ForgetContext(final ClosedContext context) {
+    final String contextId = getUniqueContextId(context);
+    final ActiveContextBridge removed = this.activeContextBridgeCache.remove(contextId);
+    if (null != removed) {
+      LOG.warning("Called with context previously unknown: " + contextId);
+    }
+  }
+
+  /**
+   * Constructs a unique ID for the given context by combining the evaluator and context ids.
+   *
+   * @param contextBase
+   * @return
+   */
+  private static String getUniqueContextId(final ContextBase contextBase) {
+    return contextBase.getEvaluatorId() + "/" + contextBase.getId();
```

Comment:
I think other parts of REEF are actually safe to use in the case of non-unique context identifiers. Hence, we have several options here if we want to keep the cache:

 1. Introduce non-deterministic behavior when the same context id is reused on the same evaluator. We could make it more deterministic by checking uniqueness in the Driver.
 2. Assemble a unique ID by walking the whole stack, concatenating the IDs of the Context in question with all its parent's IDs.
 3. Add a new, system-assigned globally unique ID to the context.

Long term, I think option 3 is the right one. For now, we could do option 2 here. Hence, I propose the following next steps:

  1. I file a JIRA for option 3.
  2. I add option 2 to the code here. That code shall be annotated with the JIRA ID of the future fix.

What do you think?